### ISSUE TITLE
pandas merge and dask merge returns different results

### ISSUE THREAD

**Issue by Pelin89** [dask/dask#7964]
Hello everyone,
I am trying to merge two dask DataFrames. However, the result is not as expected. I tried with pandas and it works fine
Pandas merge and dask merge returns different results when two DataFrame is merged.
This doesn't happen if the data is different at the early rows of the DataFrame.
Here is the code that I am using to compare two approaches.

```
import dask.dataframe as dd
import pandas as pd


df1 = dd.read_csv(file1.csv)
df2 = dd.read_csv(file2.csv)
df1_2 = df1.merge(df2, on='id')

merge with dask
dsk1 = pd.read_csv(file1.csv)
dsk2 = pd.read_csv(file2.csv)
dsk1_2 = dsk1.merge(dsk2, on='id')

```

I tested this code with these files below:
[file1.csv](https://github.com/dask/dask/files/6907802/file1.csv)
[file2.csv](https://github.com/dask/dask/files/6907803/file2.csv)

df1_2 (correct) is different than dsk1_2 (which is not correct). dsk1_2 basically merges dsk1 with itself,
which leads to a duplication and not a merge.

The first 13 rows of file1 and file2 is similar, if you change the content at first row(s) (for instance make flag1 = 10 in file1 at first row) you will notice that this time df1_2 and dsk1_2 is returning the same result.

Please let me know if dask is failing to produce the correct result when the first few rows is exactly the same 
for the two dataFrames that we are merging or if I am doing something wrong here.

Thank you!

**Comment by jsignell** [comment]
I think this is an instance of #7942 in which case it should be resolved in dask version 2021.07.2

**Comment by jsignell** [comment]
I'm going to close this, but please reopen if you still encounter this issue after upgrading to 2021.07.2
